File: app/src/main/java/com/kickstarter/viewmodels/SignupViewModel.java
package com.kickstarter.viewmodels;

import android.support.annotation.NonNull;

import com.kickstarter.libs.ActivityViewModel;
import com.kickstarter.libs.CurrentConfigType;
import com.kickstarter.libs.CurrentUserType;
import com.kickstarter.libs.Environment;
import com.kickstarter.libs.rx.transformers.Transformers;
import com.kickstarter.libs.utils.StringUtils;
import com.kickstarter.services.ApiClientType;
import com.kickstarter.services.apiresponses.AccessTokenEnvelope;
import com.kickstarter.services.apiresponses.ErrorEnvelope;
import com.kickstarter.ui.activities.SignupActivity;

import rx.Observable;
import rx.subjects.BehaviorSubject;
import rx.subjects.PublishSubject;

import static com.kickstarter.libs.rx.transformers.Transformers.takeWhen;

public interface SignupViewModel {

  interface Inputs {
    /** Call when the email field changes. */
    void email(String email);

    /** Call when the name field changes. */
    void name(String name);

    /** Call when the password field changes. */
    void password(String password);

    /** Call when the send newsletter toggle changes. */
    void sendNewslettersClick(boolean send);

    /** Call when the signup button has been clicked. */
    void signupClick();
  }

  interface Outputs {
    /** Emits a string to display when signup fails. */
    Observable<String> errorString();

    /** Emits a boolean that determines if the sign up button is enabled. */
    Observable<Boolean> formIsValid();

    /** Emits a boolean that determines if the sign up button is disabled. */
    Observable<Boolean> formSubmitting();

    /** Emits a boolean that determines if the send newsletter toggle is checked. */
    Observable<Boolean> sendNewslettersIsChecked();

    /** Finish the activity with a successful result. */
    Observable<Void> signupSuccess();
  }

  final class ViewModel extends ActivityViewModel<SignupActivity> implements Inputs, Outputs {
    private final ApiClientType client;
    private final CurrentUserType currentUser;
    private final CurrentConfigType currentConfig;

    public ViewModel(final @NonNull Environment environment) {
      super(environment);

      this.client = environment.apiClient();
      this.currentConfig = environment.currentConfig();
      this.currentUser = environment.currentUser();

      final Observable<SignupData> signupData = Observable.combineLatest(
        this.name, this.email, this.password, this.sendNewslettersIsChecked, SignupData::new
      );

      this.sendNewslettersClick
        .compose(bindToLifecycle())
        .subscribe(this.sendNewslettersIsChecked::onNext);

      signupData
        .map(SignupData::isValid)
        .compose(bindToLifecycle())
        .subscribe(this.formIsValid);

      signupData
        .compose(takeWhen(this.signupClick))
        .flatMap(this::submit)
        .compose(bindToLifecycle())
        .subscribe(this::success);

      this.currentConfig.observable()
        .take(1)
        .map(config -> false)
        .compose(bindToLifecycle())
        .subscribe(this.sendNewslettersIsChecked::onNext);

      this.signupError
        .compose(bindToLifecycle())
        .subscribe(__ -> this.koala.trackRegisterError());

      this.errorString = this.signupError
        .takeUntil(this.signupSuccess)
        .map(ErrorEnvelope::errorMessage);

      this.sendNewslettersClick
        .compose(bindToLifecycle())
        .subscribe(this.koala::trackSignupNewsletterToggle);

      this.signupSuccess
        .compose(bindToLifecycle())
        .subscribe(__ -> {
          this.koala.trackLoginSuccess();
          this.koala.trackRegisterSuccess();
        });

      this.koala.trackRegisterFormView();
    }

    private Observable<AccessTokenEnvelope> submit(final @NonNull SignupData data) {
      return this.client.signup(data.name, data.email, data.password, data.password, data.sendNewsletters)
        .compose(Transformers.pipeApiErrorsTo(this.signupError))
        .compose(Transformers.neverError())
        .doOnSubscribe(() -> this.formSubmitting.onNext(true))
        .doAfterTerminate(() -> this.formSubmitting.onNext(false));
    }

    private void success(final @NonNull AccessTokenEnvelope envelope) {
      this.currentUser.login(envelope.user(), envelope.accessToken());
      this.signupSuccess.onNext(null);
    }

    private final PublishSubject<String> email = PublishSubject.create();
    private final PublishSubject<String> name = PublishSubject.create();
    private final PublishSubject<String> password = PublishSubject.create();
    private final PublishSubject<Boolean> sendNewslettersClick = PublishSubject.create();
    private final PublishSubject<Void> signupClick = PublishSubject.create();

    private final Observable<String> errorString;
    private final PublishSubject<Void> signupSuccess = PublishSubject.create();
    private final BehaviorSubject<Boolean> formSubmitting = BehaviorSubject.create();
    private final BehaviorSubject<Boolean> formIsValid = BehaviorSubject.create();
    private final BehaviorSubject<Boolean> sendNewslettersIsChecked = BehaviorSubject.create();

    private final PublishSubject<ErrorEnvelope> signupError = PublishSubject.create();

    public final Inputs inputs = this;
    public final Outputs outputs = this;

    @Override public void email(final String email) {
      this.email.onNext(email);
    }
    @Override public void name(final String name) {
      this.name.onNext(name);
    }
    @Override public void password(final String password) {
      this.password.onNext(password);
    }
    @Override public void sendNewslettersClick(final boolean send) {
      this.sendNewslettersClick.onNext(send);
    }
    @Override public void signupClick() {
      this.signupClick.onNext(null);
    }


    @Override public @NonNull Observable<String> errorString() {
      return this.errorString;
    }
    @Override public @NonNull BehaviorSubject<Boolean> formIsValid() {
      return this.formIsValid;
    }
    @Override public @NonNull BehaviorSubject<Boolean> formSubmitting() {
      return this.formSubmitting;
    }
    @Override public @NonNull BehaviorSubject<Boolean> sendNewslettersIsChecked() {
      return this.sendNewslettersIsChecked;
    }
    @Override public @NonNull PublishSubject<Void> signupSuccess() {
      return this.signupSuccess;
    }

    final static class SignupData {
      final @NonNull String email;
      final @NonNull String name;
      final @NonNull String password;
      final boolean sendNewsletters;

      SignupData(final @NonNull String name, final @NonNull String email, final @NonNull String password,
        final boolean sendNewsletters) {
        this.name = name;
        this.email = email;
        this.password = password;
        this.sendNewsletters = sendNewsletters;
      }

      boolean isValid() {
        return this.name.length() > 0 && StringUtils.isEmail(this.email) && this.password.length() >= 6;
      }
    }
  }
}

File: app/src/main/java/com/kickstarter/viewmodels/SignupViewModel.kt
package com.kickstarter.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.kickstarter.libs.CurrentConfigTypeV2
import com.kickstarter.libs.Environment
import com.kickstarter.libs.rx.transformers.Transformers
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.isEmail
import com.kickstarter.libs.utils.extensions.isNotNull
import com.kickstarter.models.User
import com.kickstarter.services.ApiClientTypeV2
import com.kickstarter.services.apiresponses.AccessTokenEnvelope
import com.kickstarter.services.apiresponses.ErrorEnvelope
import com.kickstarter.viewmodels.usecases.LoginUseCase
import com.kickstarter.viewmodels.usecases.RefreshUserUseCase
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.PublishSubject

interface SignupViewModel {
    interface Inputs {
        /** Call when the email field changes.  */
        fun email(email: String)

        /** Call when the name field changes.  */
        fun name(name: String)

        /** Call when the password field changes.  */
        fun password(password: String)

        /** Call when the send newsletter toggle changes.  */
        fun sendNewsletters(send: Boolean)

        /** Call when the signup button has been clicked.  */
        fun signupClick()
    }

    interface Outputs {
        /** Emits a string to display when signup fails.  */
        fun errorString(): Observable<String>

        /** Emits a boolean that determines if the sign up button is disabled.  */
        fun formSubmitting(): Observable<Boolean>

        /** Finish the activity with a successful result.  */
        fun signupSuccess(): Observable<Unit>

        fun progressBarIsVisible(): Observable<Boolean>
    }

    class SignupViewModel(environment: Environment) :
        ViewModel(),
        Inputs,
        Outputs {
        private val client: ApiClientTypeV2
        private val analyticEvents = requireNotNull(environment.analytics())
        private val currentConfig: CurrentConfigTypeV2
        private val loginUserCase = LoginUseCase(environment)
        private val refreshUserUseCase = RefreshUserUseCase(environment)
        private val disposables = CompositeDisposable()

        private fun submit(data: SignupData): Observable<AccessTokenEnvelope> {
            return client.signup(
                    data.name,
                    data.email,
                    data.password,
                    data.password,
                    data.sendNewsletters
            )
                    .compose(Transformers.pipeApiErrorsToV2(signupError))
                    .compose(Transformers.neverErrorV2())
                    .doOnSubscribe {
                        this.progressBarIsVisible.onNext(true)
                        formSubmitting.onNext(true)
                    }
                    .doAfterTerminate {
                        this.progressBarIsVisible.onNext(false)
                        formSubmitting.onNext(false)
                    }
        }

        private fun success(user: User) {
            refreshUserUseCase.refresh(user)
            signupSuccess.onNext(Unit)
        }

        private val email = PublishSubject.create<String>()
        private val name = PublishSubject.create<String>()
        private val password = PublishSubject.create<String>()
        private val sendNewsletters = PublishSubject.create<Boolean>()
        private val signupClick = PublishSubject.create<Unit>()
        private val errorString: Observable<String>
        private val signupSuccess = PublishSubject.create<Unit>()
        private val formSubmitting = BehaviorSubject.create<Boolean>()
        private val signupError = PublishSubject.create<ErrorEnvelope?>()
        private val progressBarIsVisible = BehaviorSubject.create<Boolean>()

        val inputs: Inputs = this
        val outputs: Outputs = this

        init {
            client = requireNotNull(environment.apiClientV2())
            currentConfig = requireNotNull(environment.currentConfigV2())

            val signupData = Observable.combineLatest(
                    name,
                    email,
                    password,
                    sendNewsletters
            ) { name: String, email: String, password: String, sendNewsletters: Boolean ->
                SignupData(
                        name,
                        email,
                        password,
                        sendNewsletters
                )
            }

            signupData
                    .compose(Transformers.takeWhenV2(signupClick))
                    .switchMap {
                        submit(it)
                    }
                    .distinctUntilChanged()
                    .switchMap {
                        this.loginUserCase
                                .loginAndUpdateUserPrivacyV2(it.user(), it.accessToken())
                    }
                    .subscribe { success(it) }
                    .addToDisposable(disposables)

            errorString = signupError
                    .takeUntil(signupSuccess)
                    .filter { it.isNotNull() }
                    .map { it.errorMessage() }

            signupClick
                    .subscribe { analyticEvents.trackSignUpSubmitCtaClicked() }
                    .addToDisposable(disposables)

            analyticEvents.trackSignUpPageViewed()
        }

        override fun email(email: String) {
            this.email.onNext(email)
        }

        override fun name(name: String) {
            this.name.onNext(name)
        }

        override fun password(password: String) {
            this.password.onNext(password)
        }

        override fun sendNewsletters(send: Boolean) {
            sendNewsletters.onNext(send)
        }

        override fun signupClick() {
            signupClick.onNext(Unit)
        }

        override fun errorString(): Observable<String> = errorString
        override fun formSubmitting(): BehaviorSubject<Boolean> = formSubmitting
        override fun signupSuccess(): PublishSubject<Unit> = signupSuccess

        override fun progressBarIsVisible(): Observable<Boolean> = this.progressBarIsVisible

        override fun onCleared() {
            disposables.clear()
            super.onCleared()
        }

        internal class SignupData(
                val name: String,
                val email: String,
                val password: String,
                val sendNewsletters: Boolean
        )
    }

    class Factory(private val environment: Environment) : ViewModelProvider.Factory {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return SignupViewModel(environment) as T
        }
    }
}
